File: SeatingPlaceSuggestions/Java/SeatingPlaceSuggestions/src/main/java/org/weaveit/seatssuggestions/SeatingPlaceRecommender.java
package org.weaveit.seatssuggestions;

import java.util.ArrayList;
import java.util.List;

public class SeatingPlaceRecommender {
    private static final int NUMBER_OF_SUGGESTIONS = 3;
    private final AuditoriumSeatingProvider auditoriumSeatingProvider;

    public SeatingPlaceRecommender(AuditoriumSeatingProvider auditoriumLayoutAdapter) {
        this.auditoriumSeatingProvider = auditoriumLayoutAdapter;
    }

    public SuggestionsMade makeSuggestion(String showId, int partyRequested) {

        AuditoriumSeating auditoriumSeating = auditoriumSeatingProvider.findByShowId(showId);

        SuggestionsMade suggestionsMade = new SuggestionsMade(showId, partyRequested);
        
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested,
                PricingCategory.FIRST));
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating,partyRequested,
                PricingCategory.SECOND));
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating,partyRequested,
                PricingCategory.THIRD));
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating,partyRequested,
                PricingCategory.MIXED));

        if (!suggestionsMade.matchExpectations()) {
            return new SuggestionNotAvailable(showId, partyRequested);
        }

        return suggestionsMade;
    }

    private static List<SuggestionMade> giveMeSuggestionsFor(
            AuditoriumSeating auditoriumSeating, int partyRequested,  PricingCategory pricingCategory) {

        List<SuggestionMade> foundedSuggestions = new ArrayList<>();
        for (int i = 0; i < NUMBER_OF_SUGGESTIONS; i++) {
            SeatingOptionSuggested seatingOptionSuggested = auditoriumSeating.suggestSeatingOptionFor(partyRequested, pricingCategory);

            if (seatingOptionSuggested.matchExpectation()) {
                auditoriumSeating = auditoriumSeating.allocate(seatingOptionSuggested);
                foundedSuggestions.add(new SuggestionMade(seatingOptionSuggested));
            }
        }

        return List.copyOf(foundedSuggestions);
    }
}

File: SeatingPlaceSuggestions/CSharp/SeatingPlaceSuggestions/SeatingPlaceRecommender.cs
using System.Collections.Generic;

namespace SeatsSuggestions;

public class SeatingPlaceRecommender(AuditoriumSeatingAdapter auditoriumSeatingAdapter)
{
    private const int NumberOfSuggestionsPerPricingCategory = 3;

    public SuggestionsMade MakeSuggestions(string showId, int partyRequested)
    {
        var auditoriumSeating = auditoriumSeatingAdapter.FindByShowId(showId);

        var suggestionsMade = new SuggestionsMade(showId, partyRequested);

        suggestionsMade.Add(GiveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.First));
        suggestionsMade.Add(GiveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.Second));
        suggestionsMade.Add(GiveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.Third));
        suggestionsMade.Add(GiveMeSuggestionsFor(auditoriumSeating, partyRequested, PricingCategory.Mixed));

        if (suggestionsMade.MatchExpectations()) return suggestionsMade;

        return new SuggestionNotAvailable(showId, partyRequested);
    }

    private static IEnumerable<SuggestionMade> GiveMeSuggestionsFor(
        AuditoriumSeating auditoriumSeating,
        int partyRequested,
        PricingCategory pricingCategory)
    {
        var suggestionRequest = new SuggestionRequest(partyRequested, pricingCategory);
        var foundedSuggestions = new List<SuggestionMade>();

        for (var i = 0; i < NumberOfSuggestionsPerPricingCategory; i++)
        {
            var seatOptionsSuggested = auditoriumSeating.SuggestSeatingOptionFor(suggestionRequest);

            if (seatOptionsSuggested.MatchExpectation())
            {
                // We get the new version of the Auditorium after the allocation
                auditoriumSeating = auditoriumSeating.Allocate(seatOptionsSuggested);

                foundedSuggestions.Add(new SuggestionMade(seatOptionsSuggested));
            }
        }

        return foundedSuggestions;
    }
}